crates/sql/src/ast/ad_hoc.rs

```diff
@@ -0,0 +1,449 @@
+//! >  SELECT * FROM TABLE WHERE a = 'str'
+//! > ```
+//! > it cannot be evaluated when `a` is any type other than `string`.
+//!
+//! ```ebnf
+//! statement
+//!     = select
+//!     | insert
+//!     | delete
+//!     | update
+//!     | set
+//!     | show
+//!     ;
+//!
+//! insert
+//!     = INSERT INTO table [ '(' column { ',' column } ')' ] VALUES '(' literal { ',' literal } ')'
+//!     ;
+//!
+//! delete
+//!     = DELETE FROM table [ WHERE predicate ]
+//!     ;
+//!
+//! update
+//!     = UPDATE table SET [ '(' assignment { ',' assignment } ')' ] [ WHERE predicate ]
+//!     ;
+//!
+//! assignment
+//!     = column '=' expr
+//!     ;
+//!
+//! set
+//!     = SET var ( TO | '=' ) literal
+//!     ;
+//!
+//! show
+//!     = SHOW var
+//!     ;
+//!
+//! var
+//!     = ident
+//!     ;
+//!
+//! select
+//!     = SELECT [ DISTINCT ] projection FROM relation [ [ WHERE predicate ] [ ORDER BY order ] [ LIMIT limit ] ]
+//!     ;
+//!
+//! projection
+//!     = listExpr
+//!     | projExpr { ',' projExpr }
+//!     | aggrExpr { ',' aggrExpr }
+//!     ;
+//!
+//! listExpr
+//!     = STAR
+//!     | ident '.' STAR
+//!     ;
+//!
+//! projExpr
+//!     = columnExpr [ [ AS ] ident ]
+//!     ;
+//!
+//! columnExpr
+//!     = column
+//!     | qualfied
+//!     ;
+//!
+//! aggrExpr
+//!     = COUNT '(' STAR ')' [AS] ident
+//!     | COUNT '(' DISTINCT columnExpr ')' [AS] ident
+//!     | SUM   '(' columnExpr ')' [AS] ident
+//!     ;
+//!
+//! relation
+//!     = table [ [AS] ident ] { [INNER] JOIN table [ [AS] ident ] ON predicate }
+//!     ;
+//!
+//! predicate
+//!     = expr
+//!     | predicate AND predicate
+//!     | predicate OR  predicate
+//!     ;
+//!
+//! expr
+//!     = literal
+//!     | ident
+//!     | qualified
+//!     | expr op expr
+//!     ;
+//!
+//! qualified
+//!     = ident '.' ident
+//!     ;
+//!
+//! op
+//!     = '='
+//!     | '<'
+//!     | '>'
+//!     | '<' '='
+//!     | '>' '='
+//!     | '!' '='
+//!     | '<' '>'
+//!     ;
+//!
+//! order
+//!     = columnExpr [ ASC | DESC ] { ',' columnExpr [ ASC | DESC ] }
+//!     ;
+//!
+//! limit
+//!     = INTEGER
+//!     ;
+//!
+//! table
+//!     = ident
+//!     ;
+//!
+//! column
+//!     = ident
+//!     ;
+//!
+//! literal
+//!     = INTEGER
+//!     | FLOAT
+//!     | STRING
+//!     | HEX
+//!     | TRUE
+//!     | FALSE
+//!     ;
+//! ```
+use crate::ast::*;
+
+/// Arguments for `SQL` projection.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum SqlArg {
+    Column(SqlColumn),
+    QualifiedWildcard { table: Identifier },
+    Wildcard,
+}
+
+/// The `SUM` and `COUNT` expressions for `SQL` projection.
+///
+/// ```ebnf
+/// aggrExpr
+///    = COUNT '(' STAR ')' [AS] ident
+///    | COUNT '(' DISTINCT columnExpr ')' [AS] ident
+///    | SUM   '(' columnExpr ')' [AS] ident
+/// ;
+/// ```
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum SqlSummarize {
+    Count {
+        arg: SqlArg,
+        distinct: bool,
+        alias: Option<Identifier>,
```

Comment:
An Alias is required for count.

Reply:
Sorry, I meant it shouldn't be optional.

Reply:
Done